client/client-multi/client-cpp/impl/channel/KaaChannelManager.cpp

```diff
@@ -67,63 +68,112 @@ void KaaChannelManager::setFailoverStrategy(IFailoverStrategyPtr strategy) {
-                     client_->stop();
-                     break;
-                 default:
-                    break;
-             }
-        }
+        onBootstrapServerFailed(connectionInfo, reason);
     } else {
-        bootstrapManager_.useNextOperationsServer(connectionInfo->getTransportId(), reason);
+        bootstrapManager_.onOperationsServerFailed(connectionInfo->getTransportId(), reason);
     }
 }
 
+/*
+ * NOTE: Expects a valid connection info, that is not null and has the type Bootstrap.
+ */
+void KaaChannelManager::onBootstrapServerFailed(ITransportConnectionInfoPtr connectionInfo, KaaFailoverReason reason) {
+    FailoverStrategyDecision decision = failoverStrategy_->onFailover(reason);
+    switch (decision.getAction()) {
+         case FailoverStrategyAction::NOOP:
+             KAA_LOG_WARN(boost::format("No decision for Bootstrap '%s' failover") % LoggingUtils::toString(reason));
+             break;
+         case FailoverStrategyAction::RETRY_CURRENT_SERVER:
+         {
+             auto period = decision.getRetryPeriod();
+             auto bootstrapTransportId = connectionInfo->getTransportId();
+
+             KAA_LOG_WARN(boost::format("Attempt to reconnect to current Bootstrap server will be made in %d seconds") % period);
+
+             retryTimer_.stop();
+             retryTimer_.start(period, [this, bootstrapTransportId]
+                 {
+                     updateBootstrapServerAndSync(getCurrentBootstrapServer(bootstrapTransportId));
+                 });
+             break;
+         }
+         case FailoverStrategyAction::USE_NEXT_BOOTSTRAP_SERVER:
+         {
+             bool forceFirstBootstrapServer = (KaaFailoverReason::ALL_BOOTSTRAP_SERVERS_NA == reason);
+             auto nextBootstrapServer = getNextBootstrapServer(connectionInfo->getTransportId(), forceFirstBootstrapServer);
+
+             if (nextBootstrapServer) {
+                 std::size_t period = decision.getRetryPeriod();
+
+                 KAA_LOG_WARN(boost::format("Attempt to reconnect to %s Bootstrap server will be made in %d seconds")
+                                                                             % (forceFirstBootstrapServer ? "first" : "next")
+                                                                             % period);
+
+                 retryTimer_.stop();
+                 retryTimer_.start(period, [this, nextBootstrapServer]
+                     {
+                         updateBootstrapServerAndSync(nextBootstrapServer);
+                     });
+             } else {
+                 KAA_LOG_WARN(boost::format("No Bootstrap servers are accessible for %s. Processing failover...")
+                                                         % LoggingUtils::toString(connectionInfo->getTransportId()));
+
+                 onBootstrapServerFailed(connectionInfo, KaaFailoverReason::ALL_BOOTSTRAP_SERVERS_NA);
```

Comment:
Well, I could not invent something else. In which case, you think, it is a bad idea?